File: host/host.js
#!/usr/local/bin/node

// Might be good to use an explicit path to node on the shebang line in case
// it isn't in PATH when launched by Chrome.

var config = require('./config.json');
var fs = require('fs');

var nativeMessage = require('../index');

var input = new nativeMessage.Input();
var transform = new nativeMessage.Transform(messageHandler);
var output = new nativeMessage.Output();

process.stdin
    .pipe(input)
    .pipe(transform)
    .pipe(output)
    .pipe(process.stdout)
;

var subscriptions = {};

var timer = setInterval(function() {
    if (subscriptions.time) {
        output.write({ time: new Date().toISOString() });
    }
}, 1000);

input.on('end', function() {
    clearInterval(timer);
});

function messageHandler(msg, push, done) {
    
    var path = config.base_path + msg.platform + '/' + msg.title;

    if (!fs.existsSync(path)) {
		makeDirs(path);
    }
    
	fs.appendFile(path + '/tests', msg.testCases, function(err) {
		if(err) {
			return console.log(err);
		}

		console.log("The file was saved!");
	}); 

    var exec = require('child_process').exec;
    exec('gvim --remote-tab "' + path + '/' + config.program_name, function callback(error, stdout, stderr){
        //
    });
    
    
    // Just echo the message:
    push(msg.testCases);
    done();
}

function makeDirs(path) {
	var folders = path.split("/");
	if (folders.length < 1) {
		console.log("Invalid Programs path");
		return;
	}
	var currentFolder = folders[0];
	if (!fs.existsSync(currentFolder)) {
		console.log("Invalid Base Path");
		return;
	}
	for (var i = 1; i < folders.length; i++) {
		currentFolder += "/" + folders[i];
		if (!fs.existsSync(currentFolder)) {
			fs.mkdir(currentFolder);
		}
	}
}

File: host/host.py
import struct
import sys
import subprocess
import os
import json


def send_message(message):
    # Converts dictionary into string containing JSON format.
    msg_json = json.dumps(message, separators=(",", ":"))
    # Encodes string with UTF-8.
    msg_json_utf8 = msg_json.encode("utf-8")
    # Writes the message size. (Writing to buffer because writing bytes object.)
    sys.stdout.buffer.write(struct.pack("i", len(msg_json_utf8)))
    # Writes the message itself. (Writing to buffer because writing bytes object.)
    sys.stdout.buffer.write(msg_json_utf8)

# Function to read a message from Chrome.
def read_message():
    # Reads the first 4 bytes of the message (which designates message length).
    text_length_bytes = sys.stdin.buffer.read(4)

    if len(text_length_bytes) == 0:
        send_message({"error": "Empty Message"})
        sys.exit(0)

    # Unpacks the first 4 bytes that are the message length. [0] required because unpack returns tuple with required data at index 0.
    text_length = struct.unpack("i", text_length_bytes)[0]

    # Read the text (JSON object) of the message.
    text = sys.stdin.buffer.read(text_length).decode('utf-8')
    jsonRes = json.loads(text)

    # Read the config file
    with open('config-win.json') as f:
        config = json.load(f)

    # Create folder
    path = config['base_path'] + jsonRes['platform'] + '/' + jsonRes['title'] + '/'
    if not os.path.exists(path):
        os.makedirs(path)

    # Create tests file and write testCases
    if not os.path.exists(path + '/tests'):
        tests = open(path + '/tests', 'w')
        tests.write(jsonRes['tests'])
        tests.close()

    # Open Gvim
    subprocess.call(['gvim', '--remote-tab', path  + config['program_name']])

    # Program Ends Successfully
    send_message({"response": "Successfully created the tests file!"})


def Main():
    read_message()


if __name__ == '__main__':
    Main()
